pyqtorch/utils.py

```diff
@@ -116,3 +116,20 @@ def _rand(n_qubits: int) -> torch.Tensor:
 
 def param_dict(keys: Sequence[str], values: Sequence[torch.Tensor]) -> dict[str, torch.Tensor]:
     return {key: val for key, val in zip(keys, values)}
+
+
+def density_mat(
+    vector_state: torch.Tensor, dtype: torch.dtype = DEFAULT_MATRIX_DTYPE
+) -> torch.Tensor:
+    """
+    Transforms a pure state :math:`|\\psi\\rangle` into its corresponding density matrix
+    :math:`\\rho = |\\psi\\rangle \\langle\\psi|`.
+    """
+    n_qubits = len(vector_state.size()) - 1
```

Comment:
This is because in Pyq, the state size encompasses both the number of qubits and the final dimension representing the batch size. To isolate just the qubit number, we eliminate the batch dimension from the calculation using "-1".

For example:

We kwon that the two qubits state |11> is the tensor: 

         [[[0.+0.j],
         [0.+0.j]],
         [[0.+0.j],
         [1.+0.j]]]

Its size is: `torch.Size([2, 2, 1])`

But, `len(state_11.size())` = 3 
It's the number of qubits + the batch dimension

Finally,  `len(state_11.size())-1`= 2 
It's the number of qubits without the batch dimension. So just the qubit number. 
Indeed |11> is a 2 qubits state.

Reply:
OK makes sense.